app/src/main/java/org/wikipedia/watchlist/WatchlistFragment.kt

```diff
@@ -0,0 +1,309 @@
+                }))
+
+    }
+
+    private fun onSuccess(watchlistItems: List<MwQueryResult.WatchlistItem>) {
+        totalItems.clear()
+        totalItems.addAll(watchlistItems)
+
+        onUpdateList(totalItems)
+    }
+
+    private fun onError(t: Throwable) {
+        watchlistErrorView.setError(t)
+        watchlistErrorView.visibility = View.VISIBLE
+    }
+
+    private fun onUpdateList(watchlistItems: List<MwQueryResult.WatchlistItem>) {
+        val items = ArrayList<Any>()
+        items.add("") // placeholder for header
+
+        val calendar = Calendar.getInstance()
+        var curDay = -1
+
+        for (item in watchlistItems) {
+            if ((filterMode == FILTER_MODE_ALL)
+                    || (filterMode == FILTER_MODE_PAGES && Namespace.of(item.ns).main())
+                    || (filterMode == FILTER_MODE_TALK && Namespace.of(item.ns).talk())
+                    || (filterMode == FILTER_MODE_OTHER && !Namespace.of(item.ns).main() && !Namespace.of(item.ns).talk())) {
+
+                calendar.time = item.date
+                if (calendar.get(Calendar.DAY_OF_YEAR) != curDay) {
+                    curDay = calendar.get(Calendar.DAY_OF_YEAR)
+                    items.add(item.date)
+                }
+
+                items.add(item)
+            }
+        }
+
+        if (filterMode == FILTER_MODE_ALL && items.size < 2) {
+            watchlistRecyclerView.visibility = View.GONE
+            watchlistEmptyContainer.visibility = View.VISIBLE
+        } else {
+            watchlistEmptyContainer.visibility = View.GONE
+            watchlistRecyclerView.adapter = RecyclerAdapter(items)
+            watchlistRecyclerView.visibility = View.VISIBLE
+        }
+    }
+
+    internal inner class WatchlistItemViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        fun bindItem(item: MwQueryResult.WatchlistItem) {
+            val view = itemView as WatchlistItemView
+            view.setItem(item)
+            view.callback = this@WatchlistFragment
+        }
+    }
+
+    internal inner class WatchlistDateViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        fun bindItem(date: Date) {
+            val textView = itemView.findViewById<TextView>(R.id.dateText)
+            textView.text = DateUtil.getMonthOnlyDateString(date)
+        }
+    }
+
+    internal inner class WatchlistHeaderViewHolder(view: WatchlistHeaderView) : RecyclerView.ViewHolder(view) {
+        fun bindItem() {
+            (itemView as WatchlistHeaderView).callback = this@WatchlistFragment
+            itemView.enableByFilterMode(filterMode)
+        }
+    }
+
+    internal inner class RecyclerAdapter() : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+        constructor(items: List<Any>) : this() {
+            this.items = items
+        }
+
+        private var items: List<Any> = ArrayList()
+        private val VIEW_TYPE_HEADER = 0
+        private val VIEW_TYPE_DATE = 1
+        private val VIEW_TYPE_ITEM = 2
```

Comment:
Should we put these `const` variables in the `companion object`?

Reply:
yep!